### ISSUE TITLE
Expressibility boundaries: what pathmc cannot (and might) express

### ISSUE THREAD

**Issue by drbenvincent** [pymc-labs/pathmc#168]
# What pathmc cannot express (and what it might, one day)

pathmc provides a spec-driven interface for Bayesian structural causal models. You write a system of equations in a concise DSL, and pathmc compiles it to a PyMC model with a `do()` operator, identification checks, and sensitivity analysis. This is powerful for the models it targets, but like any DSL-driven tool — Bambi, brms, semopy, Stan's `brm()` — the interface necessarily restricts the class of models you can build compared to writing raw PyMC.

This document maps out those boundaries. The goal is to be honest about what falls outside pathmc's current reach, distinguish between limitations that are *accidental* (fixable with engineering effort) and those that are *essential* (outside the paradigm), and help users decide when to reach for raw PyMC instead.

## What pathmc currently supports

For context, here is the envelope of models pathmc can express today:

- **Multi-equation linear path models** with labeled coefficients, defined parameters (`:=`), and path-specific effects
- **Five likelihood families**: Gaussian, Bernoulli, Poisson, Negative Binomial, Student-t
- **Latent variables**: deterministic mediators and stochastic latent states (`latent_normal`) with sparse measurement equations
- **Residual covariance** (`~~`) via LKJ-Cholesky for Gaussian outcomes
- **Panel / longitudinal data** with partial pooling (random intercepts and slopes), lag-1 autoregressive terms, and `pytensor.scan`-compiled time dynamics
- **Transforms with estimable parameters**: geometric adstock and logistic saturation, plus a registry for custom transforms
- **Interactions** (`X:Z`, multi-way)
- **Interventional simulation** via `do()`, `ate()`, `cate()`, `att()`, `prob()`, and `simulate()`
- **Identification**: backdoor and front-door criteria, adjustment sets, collider warnings, implied conditional independence tests
- **Sensitivity analysis** for unmeasured confounding

The structural equations are **linear in coefficients** — nonlinearity enters only through transforms applied to predictors (e.g., `logistic_saturation(adstock(tv))`). The compiler maps each equation to a PyMC likelihood with a fixed link function per family (identity for Gaussian/Student-t, logit for Bernoulli, log for Poisson/NegBinomial).

## Tier 1: Natural DSL extensions

These model features are absent today but fit comfortably within pathmc's existing architecture. They require new families, link functions, or syntax — not structural changes to the compiler or `do()` engine.

### New likelihood families

| Family | What it enables | What's needed |
|--------|----------------|---------------|
| **Beta** | Rates, proportions, scores on (0,1) | Beta likelihood + logit link |
| **Gamma** | Positive continuous (durations, costs) | Gamma likelihood + log link |
| **Exponential** | Simple rate/duration models | Exponential likelihood + log link |
| **Ordinal** (cumulative) | Likert scales, ordered categories | Ordered cutpoints + cumulative logit |
| **Multinomial / Categorical** | Unordered multi-class outcomes | Softmax link, category-specific coefficients |

Each requires adding a family entry in the compiler, a default prior for any new dispersion/shape parameters, and propagation logic in `do()`. The effort is bounded and well-understood.

### Additional link functions

Currently each family has exactly one link (identity, logit, or log). Supporting a `link=` argument would enable:

- **Probit** for Bernoulli (common in econometrics)
- **Cloglog** for complementary log-log models
- **Log link for Gaussian** (log-linear models)
- **Identity link for Poisson** (linear rate models)

### Higher-order lags

The parser accepts only `lag(var)` (lag-1). Supporting `lag(var, 2)` or `lag(var, k=3)` would enable richer autoregressive dynamics. The scan compiler already handles lag-1; extending to lag-k is a matter of state vector bookkeeping, not a paradigm shift. ([#57](https://github.com/pymc-labs/pathmc/issues/57))

### More built-in transforms

Only `adstock` and `logistic_saturation` ship today. Natural additions:

- **Hill saturation** (`x^s / (x^s + k^s)`)
- **Weibull CDF saturation**
- **Power transform** (`x^alpha`)
- **Spline basis** (B-splines with estimable knot weights)

The custom transform registry already exists, so users can add these themselves. But built-in support would improve ergonomics.

### Interactions involving transforms

Currently `X:Z` interactions require plain variable names. Allowing `adstock(tv):region` or `logistic_saturation(x):Z` would broaden the spec language for MMM and moderation use cases.

## Tier 2: Significant but achievable extensions

These require meaningful engineering — new compiler pathways, extended `do()` logic, or DSL additions — but remain within pathmc's structural-equations-with-interventions paradigm.

### Zero-inflated and hurdle models

Zero-inflated Poisson/NegBinomial and hurdle models require **two linear predictors per outcome** — one for the zero-inflation probability, one for the count mean. This is a composite likelihood, not a simple family swap.

Possible DSL syntax:

```
n_purchases ~ b1*X + b2*Z            # count component
n_purchases.zi ~ g1*X                # zero-inflation component
```

Doable, but requires the compiler to emit two `mu` vectors per equation and wire them into a `pm.ZeroInflatedPoisson` (or similar). The `do()` operator would need to propagate through both components.

### Instrumental variables

pathmc can already detect IV structure from the DAG ([#114](https://github.com/pymc-labs/pathmc/issues/114)). Compiling that into a Bayesian 2SLS or full-information likelihood is the natural next step ([#115](https://github.com/pymc-labs/pathmc/issues/115)). The multi-equation compiler already handles systems; the main work is in the identification layer (recognising instruments) and adding the two-stage compilation path.

### Soft / mechanism interventions

Currently `do()` only supports hard interventions (`set X = value`). Soft interventions — scaling a coefficient, shifting a distribution, or replacing a mechanism — are planned ([#97](https://github.com/pymc-labs/pathmc/issues/97)) and would require extending `do()` to modify structural parameters rather than fixing variable values.

### Time-varying coefficients

Letting coefficients drift over time (random walks, AR(1) on betas) would capture structural change in panel models. This requires the scan compiler to treat coefficients as latent states rather than fixed draws. ([#98](https://github.com/pymc-labs/pathmc/issues/98))

### VAR(p) — multivariate autoregressive models

pathmc already handles multi-equation panel models with lag-1 terms. Extending to cross-equation lags (Y1 depends on lag of Y2) would enable Bayesian VAR models within the existing scan framework. The parser would need to accept `lag(other_var)` where `other_var` is an endogenous variable from a different equation.

### Measurement error / errors-in-variables

pathmc's latent variable support already handles the core pattern: a true variable is latent, an observed variable is a noisy measurement. Formalising this as a first-class feature (e.g., `X_obs ~ 1*X_true + measurement_noise`) with appropriate documentation would open up errors-in-variables regression. Much of the machinery exists; the gap is ergonomics and defaults.

### Non-linear structural equations

Expressions like `Y ~ b * X^2` or `Y ~ b * log(X)` are not currently parseable. The spec language is linear in coefficients with nonlinearity only via registered transforms. Supporting inline expressions (or at minimum a `transform` wrapper for common mathematical functions) would widen the expressible surface. This is a parser and compiler change, not a paradigm change.

### Difference-in-differences

DiD can already be approximated in pathmc using panel models with group × time interactions. A first-class DiD mode that enforces the parallel trends assumption, provides event-study plots, and handles staggered adoption would make this more ergonomic. The structural model underneath is still a multi-equation panel — this is a workflow convenience layer, not a fundamentally new model class.

### Panel residual covariance

The `~~` operator currently works only for cross-sectional models. Extending it to panel models (Phase 3 in the roadmap) would enable correlated residuals across equations within each time step, important for SUR-style panel models.

## Tier 3: Major rearchitecting (aspirational)

These model classes are not fundamentally incompatible with pathmc's goals, but supporting them would require substantial new infrastructure — new latent variable types, new inference strategies, or significant compiler extensions.

### Full SEM measurement models (latent factors)

pathmc supports latent *mediators* (a single latent variable determined by its structural parents), but not **latent factors** where multiple observed indicators load onto an unobserved construct:

```
# This lavaan-style syntax is NOT supported
ability =~ test1 + test2 + test3
```

Factor models require:
- A `=~` operator in the parser for measurement equations
- Factor loading matrices with identifiability constraints (fixed loadings or standardised factor variance)
- Rotation for exploratory factor analysis
- This is semopy's core territory

**Why it's hard:** The current compiler assumes every latent variable has a single structural equation determining it. Factor models invert this — the latent variable *generates* multiple indicators. The `do()` operator would also need rethinking: what does it mean to intervene on a latent factor?

**Could pathmc get here?** Possibly, for confirmatory factor analysis with a small number of pre-specified factors. Full exploratory factor analysis and IRT are likely better served by dedicated tools.

### Latent class / mixture of regressions

Models where observations belong to **discrete latent subpopulations**, each with its own regression parameters:

```python
# NOT currently expressible
# "Two types of customers with different price sensitivity"
sales ~ b1*price  # class 1
sales ~ b2*price  # class 2
```

This requires:
- A discrete latent class indicator (marginalisation or data-augmentation)
- Simplex-constrained mixing weights (Dirichlet prior)
- Label-switching mitigation
- Component-specific parameters sharing some structure

**Why it's hard:** pathmc's compiler assumes a single set of coefficients per equation. Mixture models require multiple parameter sets with probabilistic assignment. The `do()` operator would need to handle interventions within each class and marginalise across classes.

### Change-point models

Models where structural parameters shift abruptly at unknown time points. This requires discrete latent variables (change-point locations) and potentially a variable-dimension parameter space (unknown number of change-points), neither of which pathmc's compiler currently supports.

### Non-linear state-space models

pathmc's scan compiler implements a specific pattern: linear structural equations with transforms on predictors, propagated forward in time. General nonlinear state-space models (extended Kalman, unscented Kalman, particle filters) are a much broader class that would require alternative inference strategies beyond standard MCMC.

## Tier 4: Permanently out of scope

These model classes operate in fundamentally different paradigms. Supporting them would mean building a different tool, not extending pathmc.

### Gaussian processes

GPs place a prior directly on a function space, with inference over kernel hyperparameters and a Cholesky decomposition of an N×N covariance matrix (or sparse approximations). This is architecturally unrelated to systems of structural equations. PyMC has excellent GP support natively; there is no reason to route it through pathmc's DSL.

### Spatial models (CAR, ICAR, BYM2, kriging)

Spatial correlation structures require adjacency matrices, Gaussian Markov random fields, or spatial kernels. These are specialised covariance structures that don't map onto pathmc's equation-based paradigm. Libraries like PyMC's `CAR` distribution or R-INLA are the right tools.

### Survival / time-to-event models

Censoring-aware likelihoods (right, left, interval censoring), flexible baseline hazards, competing risks, and frailty models form their own modelling ecosystem. The fundamental data structure (censored event times) and likelihood construction (hazard × survival) are different from regression on fully observed outcomes.

### Bayesian neural networks / deep generative models

BNNs require high-dimensional weight-space priors, GPU-accelerated inference (SGHMC, SGLD, or variational inference), and network architecture specification. This is a fundamentally different scale and paradigm. Use PyMC + JAX, Pyro, NumPyro, or dedicated BNN libraries.

### Bayesian nonparametrics (DP mixtures, BNP density estimation)

Dirichlet Process mixtures, Pitman-Yor processes, and stick-breaking constructions require variable-dimension parameter spaces and specialised samplers. These are research-level inference problems, not something a DSL compiler should attempt.

### Causal discovery

pathmc takes a DAG as input and estimates its parameters. Learning the DAG structure from data (PC algorithm, GES, NOTEARS, etc.) is a separate problem with its own algorithms and software (causal-learn, DoWhy plugins, TETRAD). pathmc could *consume* the output of a discovery algorithm, but should not implement discovery itself.

### Copula models

Copulas separate marginal distributions from dependence structure. They require copula function evaluation, marginal CDF/quantile functions, and copula-specific parameters. This is a specialised modelling approach that doesn't fit the structural-equations paradigm.

### Quantile regression

Modelling conditional quantiles rather than conditional means requires the asymmetric Laplace likelihood (or equivalent), non-crossing constraints for simultaneous quantiles, and a different inferential target. This is a different question ("what is the 90th percentile of Y given X?") than pathmc answers ("what is E[Y | do(X=x)]?").

### Functional data analysis

When predictors or responses are *functions* (curves over a continuum), you need basis expansions, functional inner products, and smoothness penalties. This is a specialised domain with its own methodology (FDA, functional regression, functional PCA).

## The dividing line

The pattern is clear. **pathmc's scope is models expressible as a directed acyclic system of parametric structural equations** — potentially with latent variables, hierarchical structure, and nonlinear transforms — where the primary analytical goal is **causal reasoning via interventional simulation**.

A model class fits pathmc's paradigm when:
1. It can be written as a system of equations `Y_i ~ f(parents(Y_i)) + noise`
2. The causal structure is a DAG specified by the analyst
3. The primary queries are interventional: "what happens if I set X to x?"
4. Bayesian inference (MCMC/VI via PyMC) is the estimation strategy

A model class falls outside pathmc's paradigm when it requires:
- **Function-space priors** (GPs, BNP) rather than parametric equations
- **Discrete latent structure** (mixtures, HMMs, change-points) beyond what can be handled by data-augmentation within a structural equation
- **Non-DAG dependence** (spatial fields, copulas, undirected graphical models)
- **Non-standard likelihoods** (censored, functional, asymmetric) that don't decompose into per-equation terms
- **Structure learning** rather than structure-given estimation

The Tiers 1–2 items above are all cases where the model *does* fit the paradigm but pathmc's current parser or compiler doesn't yet support the necessary syntax or families. The Tier 3 items push the paradigm's boundaries. The Tier 4 items are genuinely different tools.

## Practical implication

When a user's model falls outside pathmc's envelope, the right answer is usually: **use PyMC directly** (for Bayesian models) or **use a specialised package** (semopy for CFA, CausalPy for quasi-experiments, EconML for nonparametric CATE, etc.). pathmc's `to_pymc()` method already exposes the underlying `pm.Model`, so users can extend compiled models with custom PyMC code when they need to go beyond the DSL.

The goal is not to replace PyMC — it's to make the 80% of structural causal models that *do* fit the paradigm dramatically easier to specify, fit, and query.
